### PR TITLE
Implement `Measurement` abstract class 📊

### PR DESCRIPTION
**Pull request by sduquemesa**
**Context:**
An abstract base class for measurements is implemented.

**Note**: not yet implemented for fock measurements

**Description of the Change:**
This is the diagram of the abstraction:
```mermaid
classDiagram
    Measurement <|.. Generaldyne
    Measurement <-- FockMeasurement

    class Measurement{
      +num_modes
      +postselected
      +outcome
      -_measure_gaussian(State other)
      -_measure_fock(State other)
      ~primal(State other)
      ~__lshift__(State other)
      ~__getitem__(List~int~ modes)
    }

    class Generaldyne{
      +State state
      +outcome
      ~primal(State other)
      -_measure_gaussian(State other)
      -_measure_fock(State other)
    }
    class FockMeasurement{
      (Not Integrated yet)
    }

    Generaldyne <|.. Homodyne
    Generaldyne <|.. Heterodyne
    Generaldyne *-- State

    class Homodyne{
      +State state
      -_measure_gaussian(State other)
      -_measure_fock(State other)
    }
    class Heterodyne{
      +State state
    }

    class DisplacedSqueezed{
        ...
    }
    class Coherent{
        ...
    }
    class State{
        ...
    }

    Homodyne *-- DisplacedSqueezed
    Heterodyne *-- Coherent

    FockMeasurement <|.. PNRDetector
    FockMeasurement <|.. ThresholdDetector

    class ThresholdDetector{
      (Not Integrated yet)
    }

    class PNRDetector{
      (Not Integrated yet)
    }
```

With this one can access the outcome of a measurement as
```python
detector1 = Homodyne(0.0, None)
Coherent(0.0, 0.0) << detector1
print(detector1.outcome)

detector2 = Heterodyne(None, None)
Vacuum(2) << detector2
print(detector2.outcome)

gaussian_state = Vacuum(2) >> Ggate(2)
detector3 = Generaldyne(gaussian_state)
Vacuum(2) << detector3
print(detector3.outcome)
```

- Gaussian measurements (Generaldyne, Homodyne and Heterodyne) use composition instead of inheritance for their internal state.
- Generaldyne measurements allow for projecting and sampling from any gaussian state.
- Any measurement implements the measurement and sampling logic via the private methods `_measure_fock` and `_measure_gaussian`. This is easy extensible when other representation are incorporated by implementing the new required logic.

**Benefits:**
- Given that the classes `State` and `Measurement` share very few common methods but yet their relation is necessary, composition over inheritance eases the tight coupling between states and measurements (see PR #143). In here the sampling logic is refactored so that is does not affect the `State` class.
- The above still represents well the idea that Gaussian measurements are projections onto Gaussian states. 
- The proposed abstraction creates a general framework for measurements, provides easy access to measurement outcomes when needed and respects the object flow of the operations in Mr Mustard: the `>>` operation processes `State`s and its results are always state-related — be it another `State` or a probability.

**Possible Drawbacks:**
🤷

**Related GitHub Issues:**
None

### CODE REVIEW

**Comment by ziofil** [comment]
Very nice idea, this will go a long way for extending MM.
100 points for the mermaid graph! 👍 👍

File: tests/test_lab/test_detectors.py [hunk 1]
```diff
@@ -191,62 +195,55 @@ def test_homodyne_on_2mode_squeezed_vacuum(self, s, outcome):
...
+                        + (1 + 2 * s * (1 + s)) * np.cosh(2 * r)
+                        + 2 * s * (s + 1) * np.cos(2 * angle) * np.sinh(2 * r)
+                    )
+                    / denom,
+                ],
+            ]
         )
-        assert np.allclose(remaining_state.cov, cov)
+        assert np.allclose(remaining_state.cov.numpy(), cov, atol=1e-5)
         # TODO: figure out why this is not working
```
Comment (ziofil) [inline tests/test_lab/test_detectors.py:233]: still not working?

Comment (sduquemesa) [reply tests/test_lab/test_detectors.py:233]: Nope 🫤 my suspects are units however I couldn't get it right.

File: mrmustard/lab/abstract/measurement.py [hunk 2]
```diff
@@ -15,18 +15,88 @@
...
+    @abstractmethod
+    def _measure_gaussian(self, other: State) -> Union[State, float]:
+        ...
+
+    def __lshift__(self, other) -> Union[State, float]:
+        if isinstance(other, State):
+            self.primal(other)
+        else:
+            raise TypeError(
+                f"Cannot apply Measurement '{type(self).__name__}' to '{type(other).__name__}'."
```
[the remark this answers is not in the document]
Comment (sduquemesa) [reply mrmustard/lab/abstract/measurement.py:78]: Thanks! It should be indeed
```suggestion
                f"Cannot apply Measurement '{self.__qualname__}' to '{other.__qualname__}'."
```

**Review by ziofil — APPROVED** [review]
I approve it, but let's find out what's wrong with that test

**Pull request merged**